### PR TITLE
Making `add_eos` optional in text completion dataset

### PR DESCRIPTION
**Pull request by SalmanMohammadi**
#### Context
What is the purpose of this PR? Is it to
- [x] add a new feature
- [ ] fix a bug
- [ ] update tests and/or documentation
- [ ] other (please add here)

Please link to any issues this PR addresses.

#### Changelog

Adding optional `add_eos` to text completion dataset with default set to `True`. Adding this as a seperate PR to support my PPO work https://github.com/pytorch/torchtune/pull/1005. CC @RdoubleA 

#### Test plan

- [x] run pre-commit hooks and linters (make sure you've first installed via `pre-commit install`)
- [x] add unit tests for any new functionality
- [x] update docstrings for any new or updated methods or classes
- [x] run unit tests via `pytest tests`
- [ ] run recipe tests via `pytest tests -m integration_test`
- [ ] manually run any new or modified recipes with sufficient proof of correctness
- [ ] include relevant commands and any other artifacts in this summary (pastes of loss curves, eval results, etc.)

### CODE REVIEW

File: tests/torchtune/datasets/test_text_completion_dataset.py [hunk 1]
```diff
@@ -45,3 +45,23 @@ def test_get_item(self, mock_load_dataset):
             prompt, label = dataset[i]["tokens"], dataset[i]["labels"]
             assert prompt == self.expected_tokenized_prompts[i]
             assert label == expected_labels[i]
+
+    @mock.patch("torchtune.datasets._text_completion.load_dataset")
+    def test_get_item_no_eos(self, mock_load_dataset):
+        mock_load_dataset.return_value = self.get_samples()
+        expected_labels = self.expected_tokenized_prompts
+
+        dataset = TextCompletionDataset(
```
Comment (joecummings) [inline tests/torchtune/datasets/test_text_completion_dataset.py:54]: IIUC, this test is just ensuring that a user *can* call the `TextCompletionDataset` with the param `add_eos=False`. However, we would actually want to be able to test that the EOS token is not added in this case.

**Review by joecummings — COMMENTED** [review]
Overall this looks fantastic and I think is a great value add to the API! Just one comment on how we test this :)

File: tests/torchtune/datasets/test_text_completion_dataset.py (hunk 1 above)
Comment (SalmanMohammadi) [reply tests/torchtune/datasets/test_text_completion_dataset.py:54]: apologies if I'm missing something - does`assert prompt == self.expected_tokenized_prompts[i][:-1]` not test this? I'm trimming the expected output to omit the EOS token

Comment (joecummings) [reply tests/torchtune/datasets/test_text_completion_dataset.py:54]: Ahhhh, I missed that you were trimming it to omit the EOS token - smart! Can you leave a comment in the test calling that out?

Comment (SalmanMohammadi) [reply tests/torchtune/datasets/test_text_completion_dataset.py:54]: for sure! all good now

**Review by joecummings — APPROVED** [review]
lgtm - will merge once CI is green.

**Pull request merged**